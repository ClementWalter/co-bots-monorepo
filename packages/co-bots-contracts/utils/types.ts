import { BigNumber, BigNumberish } from "ethers";
import { Collection } from "@clemlaflemme.eth/contracts/utils/types";

export type Palettes = {
  palette: Array<string>;
  collection: Collection;
};

export type Prize = {
  checkpoint: number;
  amount: BigNumberish;
  isContest: boolean;
};

export type MysteryChallenge = {
  ensId: BigNumberish;
  value: number;
  prizeIndex: number;
};

export type CoBotsParameters = {
  cobotsV1Discount: number;
  mintOutFoundersWithdrawalDelay: number;
  grandPrizeDelay: number;
  maxCobots: number;
  contestDuration: number;
  mintPublicPrice: BigNumber;
};
