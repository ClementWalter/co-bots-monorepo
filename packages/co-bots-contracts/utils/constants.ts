import fs from "fs";
import { Palettes, PalettesStorage } from "./types";

export const MAX_CONTRACT_SIZE = 24_000;
export const PALETTES_FILE =
  "../../packages/co-bots-image-processing/data/palettes.json";
export const PALETTES_ENCODED_FILE = "data/palettes-encoded.json";

export const loadPalettes = (): Palettes => {
  try {
    return JSON.parse(fs.readFileSync(PALETTES_FILE, "utf8"));
  } catch (e) {
    console.error(e);
    return {
      fill: [],
      trait: {},
      layer: [],
      layerIndexes: [],
      item: [],
    };
  }
};

export const loadPalettesEncoded = (): PalettesStorage => {
  try {
    return JSON.parse(fs.readFileSync(PALETTES_ENCODED_FILE, "utf8"));
  } catch (e) {
    console.error(e);
    return {
      fillBytes: "",
      traitBytes: "",
      traitBytesIndexes: "",
      layerIndexes: "",
    };
  }
};

// Deploy constants
export const TAGS = {
  INTEGERS: "Integers",
  CO_BOTS: "CoBots",
  CO_BOTS_PALETTES: "CoBotsPalettes",
};
