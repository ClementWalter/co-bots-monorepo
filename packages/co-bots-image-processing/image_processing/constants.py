from pathlib import Path
from string import digits

DATA_DIR = Path("data")
TRAITS_DIR = DATA_DIR / "traits"
TRAITS_COMPUTED_DIR = DATA_DIR / "traits_computed"
PALETTES_FILE = DATA_DIR / "palettes.json"

CHARACTERISTICS_ORDERED = [
    "Colour",
    "Digit 1",
    "Digit 2",
    "Digit 3",
    "Digit 4",
    "Eyes",
    "Mouth",
    "Antenna",
    "Feet",
    "Metta",
]

TRAITS_ORDERED = {
    "Colour": [
        "Black",
        "Blue",
        "Red",
    ],
    "Digit 1": list(digits),
    "Digit 2": list(digits),
    "Digit 3": list(digits),
    "Digit 4": list(digits),
    "Eyes": [
        "Classic",
        "Cyclops",
        "Awoken",
        "Flirty",
        "Zen",
        "Sadhappy",
        "Unaligned",
        "Smitten",
        "Optimistic",
        "Hacky",
        "Super",
        "Nounish",
    ],
    "Mouth": [
        "Classic",
        "Worried",
        "Knightly",
        "Shy",
        "Happy",
        "Bigsad",
        "Smug",
        "Wowed",
        "Thirsty",
        "Villainous",
        "Shady",
    ],
    "Antenna": [
        "Classic",
        "Serious",
        "Jumpy",
        "Buzzed",
        "Buggy",
        "Punk",
        "Angelic",
        "Impish",
        "Humbled",
        "Western",
        "Royal",
        "Hacky",
        "!croak",
        "⌐◨-◨",
        "Wizard",
    ],
    "Metta": ["Off", "On", "Offline"],
    "Feet": [
        "Classic",
        "Heavy Duty",
        "Firey",
        "Little Firey",
        "Roller",
        "Little Roller",
        "Energetic",
        "Little Energetic",
        "Hobbled",
        "Ghostly",
        "Pushy",
    ],
}
