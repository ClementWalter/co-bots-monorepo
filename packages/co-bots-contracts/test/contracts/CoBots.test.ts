import { setupUser, setupUsers } from "./utils";
import chai from "chai";
import {
  deployments,
  ethers,
  getNamedAccounts,
  getUnnamedAccounts,
  network,
} from "hardhat";
import { solidity } from "ethereum-waffle";
import { TAGS } from "../../utils/constants";

chai.use(solidity);
const { expect } = chai;

const setup = async () => {
  await deployments.fixture([TAGS.CO_BOTS, TAGS.CO_BOTS_PALETTES]);
  const contracts = {
    CoBots: await ethers.getContract("CoBots"),
  };
  const { deployer } = await getNamedAccounts();
  const users = await setupUsers(await getUnnamedAccounts(), contracts);
  return {
    ...contracts,
    users,
    deployer: await setupUser(deployer, contracts),
  };
};

const publicSaleFixture = deployments.createFixture(async ({ network }) => {
  const contractsAndUsers = await setup();
  const currentBlockNumber = await ethers.provider.getBlockNumber();
  const currentTimestamp = await ethers.provider
    .getBlock(currentBlockNumber)
    .then((block) => block.timestamp);
  await contractsAndUsers.deployer.CoBots.setPublicSaleTimestamp(
    currentTimestamp
  );
  await network.provider.send("evm_mine");
  return contractsAndUsers;
});

describe("CoBots", function () {
  describe("mintBatchPublicSale", async function () {
    it("should revert when minting is not open", async () => {
      const { users } = await setup();
      expect(users[0].CoBots.mintPublicSale(1)).to.be.revertedWith(
        "Public sale not open"
      );
    });
    it("should revert when price does not match", async () => {
      const { users } = await publicSaleFixture();
      expect(users[0].CoBots.mintPublicSale(1)).to.be.revertedWith(
        "Price does not match"
      );
    });
    it("should revert when quantity is too high", async () => {
      const { users } = await publicSaleFixture();
      expect(
        users[0].CoBots.mintPublicSale(21, {
          value: ethers.utils.parseEther("1.05"),
        })
      ).to.be.revertedWith(
        "Co-Bots: the requested quantity exceeds the maximum allowed"
      );
    });
    it("should mint quantity to sender", async () => {
      const { users } = await publicSaleFixture();
      await users[0].CoBots.mintPublicSale(20, {
        value: ethers.utils.parseEther("1"),
      });
      const balance = await users[0].CoBots.balanceOf(users[0].address);
      expect(balance).to.eq(20);
    });
    it("should revert when mint is closed", async () => {
      const { users } = await publicSaleFixture();
      await network.provider.send("evm_increaseTime", [168 * 60 * 60 + 1]);
      expect(
        users[0].CoBots.mintPublicSale(20, {
          value: ethers.utils.parseEther("1"),
        })
      ).to.be.revertedWith("Public sale not open");
    });
  });
});
