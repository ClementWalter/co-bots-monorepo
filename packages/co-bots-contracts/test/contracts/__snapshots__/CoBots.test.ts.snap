// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoBots mintPublicSale should mint with random seeds 1`] = `
Array [
  171,
  152,
  45,
  244,
  34,
  124,
  95,
  217,
  255,
  57,
]
`;
